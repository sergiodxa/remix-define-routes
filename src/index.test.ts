import { describe, expect, mock, test } from "bun:test";
import { defineRoutes } from ".";

// Mock node:fs to make existsSync return true always so we don't need the files
// in the test
mock.module("node:fs", () => {
	return {
		existsSync(path: string) {
			if (path.endsWith("/route.tsx") || path.endsWith("/route.ts")) {
				return false;
			}
			return true;
		},
	};
});

describe(defineRoutes.name, () => {
	let tenantRoutes = defineRoutes(({ layout }) => {
		layout(":tenant", { base: "tenant" }, ({ route }) => {
			route("/", "tenant/home");
			route("posts/:postId");
			route("users/:userId");
		});
	});

	let adminRoutes = defineRoutes(({ layout }) => {
		// Define the admin scope, move the base folder to app/admin
		layout("admin", { base: "admin" }, ({ route }) => {
			route("/", { file: "admin/home" }); // Redirect /admin to /admin/users
			route("users");
			route("posts");
		});
	});

	let debugRoutes = defineRoutes(({ base }) => {
		// Change the routes folder to app/debug
		base("debug", ({ route }) => {
			route("email");
			route("cache");
		});
	});

	let fullRoutes = defineRoutes(({ route, extend }) => {
		// Apply routes from other files or packages
		extend(tenantRoutes);
		extend(adminRoutes);
		extend(debugRoutes);
		// Overwrites /admin
		route("admin", "routes/admin/home");
		// Define / route
		route("/", "routes/home");
		route("users");
		route("posts/:postId");
		route("users/:userId", ({ route }) => {
			route("/", "routes/users.$userId._index");
			route("posts");
		});
		route("*", "routes/not-found");
	});

	test("tenant routes", () => {
		expect(tenantRoutes).toMatchSnapshot();
	});

	test("admin routes", () => {
		expect(adminRoutes).toMatchSnapshot();
	});

	test("debug routes", () => {
		expect(debugRoutes).toMatchSnapshot();
	});

	test("full routes", async () => {
		expect(fullRoutes).toMatchSnapshot();
	});
});
