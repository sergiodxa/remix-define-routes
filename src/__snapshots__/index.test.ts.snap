// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`defineRoutes tenant routes 1`] = `
{
  "routes/$tenant": {
    "file": "tenant/home.tsx",
    "id": "routes/$tenant",
    "index": false,
    "parentId": "routes/$tenant",
    "path": "/",
  },
  "routes/$tenant.posts.$postId": {
    "file": "routes/$tenant.posts.$postId.tsx",
    "id": "routes/$tenant.posts.$postId",
    "index": false,
    "parentId": "routes/$tenant",
    "path": "posts/:postId",
  },
  "routes/$tenant.users.$userId": {
    "file": "routes/$tenant.users.$userId.tsx",
    "id": "routes/$tenant.users.$userId",
    "index": false,
    "parentId": "routes/$tenant",
    "path": "users/:userId",
  },
}
`;

exports[`defineRoutes admin routes 1`] = `
{
  "routes/admin": {
    "file": "admin/home.tsx",
    "id": "routes/admin",
    "index": false,
    "parentId": "routes/admin",
    "path": "/",
  },
  "routes/admin.posts": {
    "file": "routes/admin.posts.tsx",
    "id": "routes/admin.posts",
    "index": false,
    "parentId": "routes/admin",
    "path": "posts",
  },
  "routes/admin.users": {
    "file": "routes/admin.users.tsx",
    "id": "routes/admin.users",
    "index": false,
    "parentId": "routes/admin",
    "path": "users",
  },
}
`;

exports[`defineRoutes debug routes 1`] = `
{
  "debug/cache": {
    "file": "debug/cache.tsx",
    "id": "debug/cache",
    "index": false,
    "parentId": "root",
    "path": "cache",
  },
  "debug/email": {
    "file": "debug/email.tsx",
    "id": "debug/email",
    "index": false,
    "parentId": "root",
    "path": "email",
  },
}
`;

exports[`defineRoutes full routes 1`] = `
{
  "debug/cache": {
    "file": "debug/cache.tsx",
    "id": "debug/cache",
    "index": false,
    "parentId": "root",
    "path": "cache",
  },
  "debug/email": {
    "file": "debug/email.tsx",
    "id": "debug/email",
    "index": false,
    "parentId": "root",
    "path": "email",
  },
  "routes": {
    "file": "routes/home.tsx",
    "id": "routes",
    "index": false,
    "parentId": "root",
    "path": "/",
  },
  "routes/$tenant": {
    "file": "tenant/home.tsx",
    "id": "routes/$tenant",
    "index": false,
    "parentId": "routes/$tenant",
    "path": "/",
  },
  "routes/$tenant.posts.$postId": {
    "file": "routes/$tenant.posts.$postId.tsx",
    "id": "routes/$tenant.posts.$postId",
    "index": false,
    "parentId": "routes/$tenant",
    "path": "posts/:postId",
  },
  "routes/$tenant.users.$userId": {
    "file": "routes/$tenant.users.$userId.tsx",
    "id": "routes/$tenant.users.$userId",
    "index": false,
    "parentId": "routes/$tenant",
    "path": "users/:userId",
  },
  "routes/*": {
    "file": "routes/not-found.tsx",
    "id": "routes/*",
    "index": false,
    "parentId": "root",
    "path": "*",
  },
  "routes/admin": {
    "file": "routes/admin/home.tsx",
    "id": "routes/admin",
    "index": false,
    "parentId": "root",
    "path": "admin",
  },
  "routes/admin.posts": {
    "file": "routes/admin.posts.tsx",
    "id": "routes/admin.posts",
    "index": false,
    "parentId": "routes/admin",
    "path": "posts",
  },
  "routes/admin.users": {
    "file": "routes/admin.users.tsx",
    "id": "routes/admin.users",
    "index": false,
    "parentId": "routes/admin",
    "path": "users",
  },
  "routes/posts.$postId": {
    "file": "routes/posts.$postId.tsx",
    "id": "routes/posts.$postId",
    "index": false,
    "parentId": "root",
    "path": "posts/:postId",
  },
  "routes/users": {
    "file": "routes/users.tsx",
    "id": "routes/users",
    "index": false,
    "parentId": "root",
    "path": "users",
  },
  "routes/users.$userId": {
    "file": "routes/users.$userId._index.tsx",
    "id": "routes/users.$userId",
    "index": false,
    "parentId": "routes/users.$userId",
    "path": "/",
  },
  "routes/users.$userId.posts": {
    "file": "routes/users.$userId.posts.tsx",
    "id": "routes/users.$userId.posts",
    "index": false,
    "parentId": "routes/users.$userId",
    "path": "posts",
  },
}
`;
